src/middlewared/middlewared/plugins/docker/state_setup.py

```diff
@@ -0,0 +1,69 @@
+import os
+import shutil
+import uuid
+
+from datetime import datetime
+
+from middlewared.service import private, Service
+
+from .state_utils import DATASET_DEFAULTS, docker_datasets, docker_dataset_custom_props, docker_dataset_update_props
+from .utils import applications_ds_name
+
+
+class DockerSetupService(Service):
+
+    class Config:
+        namespace = 'docker.setup'
+        private = True
+
+    @private
+    async def status_change(self):
+        config = await self.middleware.call('docker.config')
+        if not config['pool']:
+            return
+
+        await self.pools(config)
+
+    @private
+    async def pools(self, config):
+        return await self.create_update_docker_datasets(applications_ds_name(config['pool']))
+
+    @private
+    async def create_update_docker_datasets(self, docker_ds):
+        create_props_default = DATASET_DEFAULTS.copy()
+        for dataset_name in docker_datasets(docker_ds):
+            custom_props = docker_dataset_custom_props(dataset_name.split('/', 1)[-1])
+            # got custom properties, need to re-calculate
+            # the update and create props.
+            create_props = dict(create_props_default, **custom_props) if custom_props else create_props_default
+            update_props = docker_dataset_update_props(create_props)
+
+            dataset = await self.middleware.call(
+                'zfs.dataset.query', [['id', '=', dataset_name]], {
+                    'extra': {
+                        'properties': list(update_props),
+                        'retrieve_children': False,
+                        'user_properties': False,
+                    }
+                }
+            )
+            if not dataset:
+                test_path = os.path.join('/mnt', dataset_name)
+                if await self.middleware.run_in_thread(os.path.exists, test_path):
```

Comment:
You're creating 2 threads everytime this method is called to perform a `shutil.move` which is unnecessary.

This is prone to TOCTOU issues as well so you can simply do `shutil.move` and catch `FileNotFoundError` since it'll fail if `test_path` doesn't exist.